src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -115,6 +130,247 @@ public static Task<List<string>> ReadRequestAndSendResponseAsync(Socket server,
+                case AuthenticationProtocols.Basic:
+                    success = IsBasicAuthTokenValid(line, options);
+                    break;
+
+                case AuthenticationProtocols.Digest:
+                    // Read the request content.
+                    string requestContent = null;
+                    while (!string.IsNullOrEmpty(line = await reader.ReadLineAsync().ConfigureAwait(false)))
+                    {
+                        if (line.Contains("Content-Length"))
+                        {
+                            line = await reader.ReadLineAsync().ConfigureAwait(false);
+                            while (!string.IsNullOrEmpty(line = await reader.ReadLineAsync().ConfigureAwait(false)))
+                            {
+                                requestContent += line;
+                            }
+                        }
+                    }
+
+                    success = IsDigestAuthTokenValid(clientResponse, requestContent, requestMethod, options);
+                    break;
+            }
+
+            if (success)
+            {
+                await writer.WriteAsync(DefaultHttpResponse).ConfigureAwait(false);
+            }
+            else
+            {
+                await writer.WriteAsync(response).ConfigureAwait(false);
+            }
+
+            return null;
+        }
+
+        private static bool IsBasicAuthTokenValid(string clientResponse, Options options)
+        {
+            string clientHash = clientResponse.Substring(clientResponse.IndexOf(nameof(AuthenticationProtocols.Basic), StringComparison.OrdinalIgnoreCase) +
+                nameof(AuthenticationProtocols.Basic).Length).Trim();
+            string userPass = string.IsNullOrEmpty(options.Domain) ? options.Username + ":" + options.Password : options.Domain + "\\" + options.Username + ":" + options.Password;
+            return clientHash == Convert.ToBase64String(Encoding.UTF8.GetBytes(userPass));
+        }
+
+        private static bool IsDigestAuthTokenValid(string clientResponse, string requestContent, string requestMethod, Options options)
+        {
+            string clientHash = clientResponse.Substring(clientResponse.IndexOf(nameof(AuthenticationProtocols.Digest), StringComparison.OrdinalIgnoreCase) +
+                nameof(AuthenticationProtocols.Digest).Length).Trim();
+            string[] values = clientHash.Split(',');
+
+            string username = null, uri = null, realm = null, nonce = null, response = null, algorithm = null, cnonce = null, opaque = null, qop = null, nc = null;
+            bool userhash = false;
+            for (int i = 0; i < values.Length; i++)
+            {
+                string trimmedValue = values[i].Trim();
+                if (trimmedValue.Contains(nameof(username)))
+                {
+                    // Username is a quoted string.
+                    int startIndex = trimmedValue.IndexOf('"') + 1;
+
+                    if (startIndex != -1)
```

Comment:
I don't understand this logic.  In what circumstance will startIndex ever be -1?  For that to happen, IndexOf would need to be returning -2, which it will never do.

Same comment for all other occurrences of this below.

Reply:
Oh that's old code, I forgot to change this when i made startindex point to the valid char to be taken.